### PR TITLE
Object array traversal

### PR DESCRIPTION
**Pull request by rase-**
Ordering, filtering and traversal support for array type with nested objects.

### CODE REVIEW

File: server/modules/selva/module/find.c [hunk 1]
```diff
@@ -864,6 +983,183 @@ static int send_node_fields(RedisModuleCtx *ctx, RedisModuleString *lang, SelvaM
...
+     *       ...
+     *       fieldNameN,
+     *       fieldValueN,
+     *     ]
+     *   ]
+     * ```
+     */
+
+    RedisModule_ReplyWithArray(ctx, 2);
+    RedisModule_ReplyWithString(ctx, RedisModule_CreateString(ctx, EMPTY_NODE_ID, SELVA_NODE_ID_SIZE));
```
Comment (OlliV) [inline server/modules/selva/module/find.c:1120]:
```suggestion
    RedisModule_ReplyWithStringBuffer(ctx, EMPTY_NODE_ID, SELVA_NODE_ID_SIZE);
```

File: server/modules/selva/module/find.c [hunk 2]
```diff
@@ -1304,6 +1600,85 @@ static int FindCommand_NodeCb(struct SelvaModify_HierarchyNode *node, void *arg)
...
+    if (take) {
+        const int sort = !!args->order_field;
+
+        if (!sort) {
+            ssize_t *nr_nodes = args->nr_nodes;
+            ssize_t * restrict limit = args->limit;
+            int err;
+
+            if (args->fields) {
+                fprintf(stderr, "HELLO SENDING FIELDS FOR OBJ\n");
```
Comment (OlliV) [inline server/modules/selva/module/find.c:1642]:
```suggestion
```

File: server/modules/selva/module/find.c [hunk 3]
```diff
@@ -1304,6 +1600,85 @@ static int FindCommand_NodeCb(struct SelvaModify_HierarchyNode *node, void *arg)
...
+static int FindCommand_ArrayNodeCb(struct SelvaObject *obj, void *arg) {
+    fprintf(stderr, "HELLO ARRAY NODE CB\n");
+    struct FindCommand_Args *args = (struct FindCommand_Args *)arg;
+    struct rpn_ctx *rpn_ctx = args->rpn_ctx;
+    int take = (args->offset > 0) ? !args->offset-- : 1;
+
+    if (take && rpn_ctx) {
+        int err;
+
+        /* Set node_id to the register */
```
Comment (OlliV) [inline server/modules/selva/module/find.c:1612]:
```suggestion
        /* Set obj to the register */
```

File: server/modules/selva/module/find.c [hunk 4]
```diff
@@ -1304,6 +1600,85 @@ static int FindCommand_NodeCb(struct SelvaModify_HierarchyNode *node, void *arg)
     return 0;
 }
 
+static int FindCommand_ArrayNodeCb(struct SelvaObject *obj, void *arg) {
+    fprintf(stderr, "HELLO ARRAY NODE CB\n");
```
Comment (OlliV) [inline server/modules/selva/module/find.c:1604]:
```suggestion
```

File: server/modules/selva/module/find.c [hunk 5]
```diff
@@ -1765,7 +2194,14 @@ static int SelvaHierarchy_Find(RedisModuleCtx *ctx, int recursive, RedisModuleSt
 
             err = SelvaModify_TraverseHierarchyRef(ctx, hierarchy, nodeId, ref_field_str, &cb);
         } else if (dir == SELVA_HIERARCHY_TRAVERSAL_ARRAY && ref_field) {
-            // err = SelvaModify_TraverseArray(ctx, hierarchy, nodeId, ref_field_str, &cb);
+            TO_STR(ref_field);
+            array_traversal_ref_field = ref_field_str;
+            fprintf(stderr, "HELLO ARRAY TRAVERSAL YOYO %.*s\n", (int)ref_field_len, ref_field_str);
```
Comment (OlliV) [inline server/modules/selva/module/find.c:2199]:
```suggestion
```

File: server/modules/selva/module/find.c [hunk 6]
```diff
@@ -864,6 +983,183 @@ static int send_node_fields(RedisModuleCtx *ctx, RedisModuleString *lang, SelvaM
...
+    if (err) {
+        fprintf(stderr, "%s:%d: Failed to send the field (%s) for node_id: \"%.*s\" err: \"%s\"\n",
+                __FILE__, __LINE__,
+                field_str,
+                (int)SELVA_NODE_ID_SIZE, EMPTY_NODE_ID,
+                getSelvaErrorStr(err));
+        RedisModule_ReplyWithNull(ctx);
+    }
+
+    return 1;
```
Comment (OlliV) [inline server/modules/selva/module/find.c:1094]:
```suggestion
    return 1;
#undef SEND_FIELD_NAME
```

**Review by OlliV — CHANGES_REQUESTED** [review]

File: client/test/getBasic.ts [hunk 7]
```diff
@@ -2275,49 +2275,6 @@ test.serial('get - record with nested wildcard query', async (t) => {
 
 test.serial('get - field with array', async (t) => {
   const client = connect({ port })
-
```
[the remark this answers is not in the document]
Comment (rase-) [reply client/test/getBasic.ts:2278]: Still using this for debugging very often so I suggest cleaning it up later. It's commented out so it's fine.

**Review by OlliV — APPROVED** [review]

**Pull request merged**